dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/SharingUtils.java

```diff
@@ -51,17 +49,64 @@ private SharingUtils()
         throw new UnsupportedOperationException( "utility" );
     }
 
-    public static Set<UserGroupAccess> getDtoUserGroupAccesses( Sharing sharing )
+    public static Set<UserGroupAccess> getDtoUserGroupAccesses( Set<org.hisp.dhis.user.UserGroupAccess> dto,
+        Sharing sharing )
     {
-        return sharing.hasUserGroupAccesses() ? sharing.getUserGroups().values()
-            .stream().map( org.hisp.dhis.user.sharing.UserGroupAccess::toDtoObject ).collect( Collectors.toSet() )
-            : new HashSet<>();
+        if ( !CollectionUtils.isEmpty( dto ) )
+        {
+            return dto;
+        }
+
+        if ( dto == null )
+        {
+            dto = new HashSet<>();
+        }
+
+        if ( sharing.hasUserGroupAccesses() )
+        {
+            dto = new HashSet<>( sharing.getUserGroups().values()
```

Comment:
Unnecessary to wrap/copy the result of stream `collect` to `HashSet` again.

Reply:
fixed